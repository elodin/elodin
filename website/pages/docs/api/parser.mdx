import DocLayout from '../../../components/DocLayout.js'

export default DocLayout

# Parser

```bash
yarn add --dev @elodin/parser
```

The parser turns a string of Elodin code into an AST (Abstract Syntax Tree).

## parse

It exposes a `parse` method that just takes an single string of code.

### Example

```javascript
import { parse } from '@elodin/parser'

const input = `
style Button {
  backgroundColor: red
  paddingLeft: 10
  paddingRight: 10
}
`

const ast = parse(input)
```

## AST Nodes

The AST consists of the following AST nodes:

### File

| Property | Type             | Description        |
| -------- | ---------------- | ------------------ |
| body     | [Variant, Style] | List of root nodes |

<br />
<br />

### Style

| Property | Type                       | Description                                  |
| -------- | -------------------------- | -------------------------------------------- |
| name     | String                     | The name of the style                        |
| format   | `view`, `text`             | The format of the style                      |
| body     | [Declaration, Conditional] | List of declarations                         |
| comments | [String]                   | List of comments found in front of the style |

<br />
<br />

### Variant

| Property | Type         | Description                                  |
| -------- | ------------ | -------------------------------------------- |
| name     | String       | The name of the variant                      |
| body     | [Identifier] | List of variant values                       |
| comments | [String]     | List of comments found in front of the style |

<br />
<br />

### Conditional

| Property | Type                               | Description                                                                                            |
| -------- | ---------------------------------- | ------------------------------------------------------------------------------------------------------ |
| property | String                             | The property that is being compared (only available if boolean is `false`)                             |
| value    | Float, Integer, Identifier, String | The value that is being compared with the actual property value (only available if boolean is `false`) |
| operator | `=`, `>=`, `>`, `<=`, `<`          | The comparison operator used                                                                           |
| boolean  | Boolean                            | Whether the condition is boolean                                                                       |
| body     | [Declaration, Conditional]         | List of declarations                                                                                   |

<br />
<br />

### Declaration

| Property | Type                                                               | Description                                                           |
| -------- | ------------------------------------------------------------------ | --------------------------------------------------------------------- |
| property | String                                                             | The style property                                                    |
| value    | [FunctionExpression, Float, Integer, Identifier, String, Variable] | The style value                                                       |
| dynamic  | Boolean                                                            | Whether it's declaration has dynamic input (thus a variable as value) |
| raw      | Boolean                                                            | Whether it has a raw property prefixed with `__`                      |

<br />
<br />

### FunctionExpression

| Property | Type | Description |
| -------- | ---- | ----------- |
| callee   |      |             |
| params   |      |             |

<br />
<br />

### Float

| Property   | Type | Description |
| ---------- | ---- | ----------- |
| integer    |      |             |
| fractional |      |             |
| negative   |      |             |

<br />
<br />

### Integer

| Property | Type | Description |
| -------- | ---- | ----------- |
| value    |      |             |
| negative |      |             |

<br />
<br />

### Identifier

| Property | Type | Description |
| -------- | ---- | ----------- |
| value    |      |             |

<br />
<br />

### String

| Property | Type | Description |
| -------- | ---- | ----------- |
| value    |      |             |

<br />
<br />

### Variable

| Property | Type | Description |
| -------- | ---- | ----------- |
| value    |      |             |
