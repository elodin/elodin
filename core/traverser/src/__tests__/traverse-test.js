import { parse } from '@elodin/parser'

import traverse from '../traverse'

describe('Traversing ast nodes', () => {
  it('should correctly traverse nodes', () => {
    const file = `
style Button {
  color: red
  [Type=Primary] {
    color: blue
  }
}`

    const log = []

    const visitor = {
      Identifier: {
        enter(path) {
          log.push(path.node.value)
          path.node.value = path.node.value.toUpperCase()
          log.push(path.node.value)
        },

        exit(path) {
          log.push(path.node.value)
        },
      },
    }

    const { ast } = parse(file)

    expect(traverse(ast, [visitor])).toMatchSnapshot()
    expect(log).toMatchSnapshot()
  })
})
