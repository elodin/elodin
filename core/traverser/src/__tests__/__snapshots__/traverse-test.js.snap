// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Traversing ast nodes should correctly traverse nodes 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "comments": Array [],
          "dynamic": false,
          "property": "color",
          "raw": false,
          "type": "Declaration",
          "value": Object {
            "type": "Identifier",
            "value": "RED",
          },
        },
        Object {
          "body": Array [
            Object {
              "comments": Array [],
              "dynamic": false,
              "property": "color",
              "raw": false,
              "type": "Declaration",
              "value": Object {
                "type": "Identifier",
                "value": "BLUE",
              },
            },
          ],
          "comments": Array [],
          "operator": "=",
          "property": Object {
            "type": "Identifier",
            "value": "TYPE",
          },
          "type": "Conditional",
          "value": Object {
            "type": "Identifier",
            "value": "PRIMARY",
          },
        },
      ],
      "comments": Array [],
      "name": "Button",
      "type": "Style",
    },
  ],
  "type": "File",
}
`;

exports[`Traversing ast nodes should correctly traverse nodes 2`] = `
Array [
  "red",
  "RED",
  "RED",
  "Primary",
  "PRIMARY",
  "PRIMARY",
  "Type",
  "TYPE",
  "TYPE",
  "blue",
  "BLUE",
  "BLUE",
]
`;
